feat!: restructure folders

joachimdekker opened the pull request:
## Description

This PR proposes to change the folder structure of aplib. It puts GoalStructure in its own folder, and creates a seperate namespace for agent.

BREAKING CHANGE: Because agent has a seperate namespace, this PR introduces a breaking change.

Closes AP-117.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch

joachimdekker committed:
feat!: fix filestructure

joachimdekker committed:
refactor: rearrange Goal class elements and improve formatting

The elements within the Goal class have been rearranged for better readability. The epsilon variable and the heuristic function, used to determine if a goal has been met, have been moved down below. Additionally, some formatting has been performed to make the constructor overloads more readable.

SilasPeters commented on Aplib.Core/Desire/Goals/Goal.cs:
```diff
@@ -6,28 +6,15 @@ namespace Aplib.Core.Desire.Goals
-    /// the testing.
+    /// applying the tactic. Goals are combined in a <see cref="GoalStructure.GoalStructure{TBeliefSet}" />, and are used to
+    /// prepare tests
+    /// or do the testing.
```
grrrrr fix this :(((((((((((

joachimdekker replied:
Rider won't let me :(

QuakeEye replied:
common Rider L

joachimdekker committed:
chore: fix comments in Goal.cs

The comments in the Goal.cs file have been reformatted for better readability. The description about the usage of Goals has been merged into one single line.

SilasPeters reviewed (dismissed):
You did not give any formal proofs. However, I wouldn't even check them if you did.

JensSteenmetz commented on Aplib.Tests/Desire/GoalStructureTests.cs:
```diff
@@ -1,9 +1,11 @@
 using FluentAssertions;
 using Moq;
 using Moq.Protected;
+using System;
```
Nit: could this using statement be removed? I would like to prevent using System namespace in aplib code, since the System.Action delegate clashes with Aplib.Core.Intent.Actions.Action. Even though Aplib.Core.Intent.Actions is not used in this script currently, it might be used later.

joachimdekker replied:
This is the test script. Is that really necessary?

JensSteenmetz replied:
It is not giving issues right now, but it's sort of personal preference. I don't like having using directives that could potentially cause issues down the line. But not necessary, hence the nit. So you can leave it if you disagree.

JensSteenmetz commented on Aplib.Core/Desire/Goals/Goal.cs:
Could this folder/namespace be renamed to Goal? To be consistent with the GoalStructure namespace? Same comment on Aplib.Core.Intent.Actions and Aplib.Core.Intent.Tactics

joachimdekker replied:
I don't agree. It houses multiple `Goals` just like the `System.Collections`. I will change the `GoalStructure` folder to `GoalStructures`.

JensSteenmetz replied:
That's also fine

JensSteenmetz reviewed (commented):
Could you also introduce DesireSet, Belief.Belief and Belief.BeliefSet namespaces?

JensSteenmetz reviewed (changes requested):
Left a Comment instead of Request changes by accident.

JensSteenmetz commented:
To formally prove the correctness of our new folder structure, we introduce a formalized set-theoretic approach. Let's define our folder structure as a set $F$ and introduce a function $\phi$ that maps elements of $F$ to their respective responsibilities within the system.

**Axiom 1: Modularity (M)**
A folder structure $F$ is said to be modular if for every element $f \in F$, there exists a unique responsibility $r$ such that $\phi(f) = r$. This implies a one-to-one correspondence between folders and responsibilities.

**Axiom 2: Scalability (S)**
A folder structure $F$ is scalable if for any non-empty subset $G \subseteq F$, there exists an element $g \notin F$ such that $F \cup \{g\}$ is also a folder structure satisfying Axiom 1.

**Theorem: The folder structure $F$ is effective if it satisfies both Modularity (M) and Scalability (S).**

**Proof:**

Let $F = \{ Core, Agent, Belief, BeliefSet, Desire, GoalStructure, Goal, DesireSet, Intent, Action, Tactic, Collections \}$.

**Step 1: Verify Modularity**
Define $\phi: F \rightarrow R$ where $R$ is the set of responsibilities. For instance, $\phi(Agent)$ could be 'Agent Operations', $\phi(Belief)$ could be 'Belief Management', and so on.

Since each folder in $F$ has a distinct responsibility, and no two folders share the same responsibility, $\phi$ is injective, thus satisfying Modularity (M).

**Step 2: Verify Scalability**
Consider a subset $G \subseteq F$, say $G = \{ Belief, BeliefSet \}$. If we want to add a new belief $b$, we can do so by creating a new element $g \notin F$, such that $g = b$ and $\phi(g) = 'New Belief'$.

Since adding $g$ to $F$ does not require changing the existing structure and $\phi$ remains injective, $F$ satisfies Scalability (S).

By demonstrating that $F$ satisfies both Modularity (M) and Scalability (S), we conclude that the folder structure is effective according to our theorem.

joachimdekker committed:
feat: conform to Jens

SilasPeters reviewed (changes requested):
The Aplib.Tests project should be renamed to Aplib.Core.Tests, for every project needs its own tests project

JensSteenmetz commented on Aplib.Core/Agents/BdiAgent.cs:
```diff
@@ -4,14 +4,17 @@
     public class BdiAgent<TBeliefSet> : IAgent
         where TBeliefSet : IBeliefSet
     {
+        /// <inheritdoc />
+        public CompletionStatus Status => _desireSet.Status;
+
```
Read-only should be above non-read-only public fields and properties

JensSteenmetz reviewed (changes requested):
One small thing regarding our conventions, lgtm otherwise

joachimdekker committed:
chore: move files to different folder.

joachimdekker committed:
merge: dev

joachimdekker committed:
feat: move new tests to new test folder

joachimdekker committed:
chore: use new test project location

joachimdekker committed:
chore: fix some merge issues

joachimdekker committed:
revert: stupid change

joachimdekker committed:
feat: add more seperate namespaces

JensSteenmetz reviewed (approved):
LGTMMMM
